### ISSUE TITLE
(tracking issue) Most important Zig issues affecting Nuitka

### ISSUE THREAD

**Issue by jimkring** [Nuitka/Nuitka#2742]
> Note: This issue was created simply to provide a way for us (Nuitka project) to communicate to the Zig project about our most important Zig issues. We have provided them this link on their [Third Party Tracking Issues (what is important to other people?)](https://github.com/ziglang/zig/wiki/Third-Party-Tracking-Issues-(what-is-important-to-other-people%3F)) page

Nuitka's wish list for the Zig project!

Currently, we are trying to [support Zig as a backend C compiler for Nuitka](https://github.com/Nuitka/Nuitka/issues/2741). 

Some specific pain points, from highest priority to lowest priority:

-  https://github.com/ziglang/translate-c/issues/149
- https://github.com/ziglang/zig/issues/8973 (we can probably work around this by adding special support for Zig)

**Comment by jimkring** [comment]
@kayhayen FYI, this “issue” #2742 is not really a Nuitka issue/feature/bug. It’s simply a place for us to communicate to the Zig team what we need from them.

**Comment by polarathene** [comment]
@jimkring I thin you accidentally linked to a Zig issue when intending to refer to this Nuitka issue number.

---

@kayhayen Rather than going off-topic at another issue I am responding here.

From: https://github.com/Nuitka/Nuitka-website/issues/124#issuecomment-3418685347

> _Currently zig works only via the C++03 workaround, which limits its use to Python3.10 as we need full C11 support with atomics_
> _It's major strength would be cross compilation, but Nuitka itself needs to learn that properly first_

We have already recently discussed at that `translate-c` issue that it's not related to `zig cc` / `zig c++`. Do you have a code snippet that fails to compile with Zig that I could use please?

---

I have successfully compiled Zig with Python 3.11 without any issues thus far. I've also patched a few lines (_in generated C, I've not searched for equivalent logic in Nuitka source_) to get Python 3.12 and 3.13 working too.

I can also confirm success in cross-compilation from x86_64 to aarch64 and any desired glibc version target (_for broader compatibility without needing a legacy/EOL build environment_). This is all handled via two scripts.
- A `CC` wrapper script to modify the invocations from Nuitka (_and also from UV when building packages to install from sdist if needing to lower glibc target_).
- A cross-compilation support script that ensures Nuitka is given `-I` + `-L` flags for precedence of the compile target arch, and then copies over the correct libpython.so to link against. I've not yet tried cross compiling to Windows or macOS platforms.

That said it is not entirely perfect. Notably when building packages via sdists, some provide a linker flag that Zig does not support, [`--exclude-libs`](https://github.com/ziglang/zig/issues/15113).

---

Related is cross-compiling to WASM/WASI, the `wasm32-wasi` target fails [due to some missing headers](https://github.com/Nuitka/Nuitka/issues/2741#issuecomment-3404019844), which I believe there has been some work towards with Nuitka, but I am unclear of the status on that and if Nuitka is able to provision these yet.

**Comment by kayhayen** [comment]
I will need to verify by downloading and attempting latest zig, but actual zig didn't support C11 atomics when I last tried it. I was mistakenly assuming the wrong issue to be about adding support for that. So I might have missed that it was added to `zig cc` in the mean time. I am checking that now.

**Comment by kayhayen** [comment]
So, I managed to get it to work on Linux at least, which is great progress. It's a shame we didn't notice that it started to work. I am using zig 0.15.2 and will put it on our factory branch and track things from there. Strangely linking didn't work for me any longer, as it also needs a "cc" injection. I believe I had it once working for older Python, but as usual I am now convinced that can be the case.

There were a few warnings for generated code to white list. When I run it in debug mode, I get a panic, that seems unwarranted:

```
Nuitka: Launching 'AssertsTest.bin'.
thread 358117 panic: null pointer passed as argument 2, which is declared to never be null
static_src/CompiledFunctionType.c:1392:5: 0x1084e0b in Nuitka_Function_New (static_src/CompiledFunctionType.c)
/home/hayen/repos/Py2C/AssertsTest.build/module.__main__.c:661:44: 0x1076284 in MAKE_FUNCTION___main__$$36$$$36$$$36$function__1_testAssert1 (module.__main__.c)
    struct Nuitka_FunctionObject *result = Nuitka_Function_New(
                                           ^
/home/hayen/repos/Py2C/AssertsTest.build/module.__main__.c:1138:31: 0x106e9e8 in modulecode___main__ (module.__main__.c)
        tmp_assign_source_6 = MAKE_FUNCTION___main__$$36$$$36$$$36$function__1_testAssert1(tstate);
```

My C code doesn't make a promise of argument 2 to not be NULL, nor do I believe it would be, unless some other bug occurs of course, we are not using `__nonnull__` or anything like that. The relevant declaration is this:

```
struct Nuitka_FunctionObject *Nuitka_Function_New(function_impl_code c_code, PyObject *name, PyObject *qualname,
                                                  PyCodeObject *code_object, PyObject *defaults, PyObject *kw_defaults,
                                                  PyObject *annotations, PyObject *module, PyObject *doc,
                                                  struct Nuitka_CellObject **closure, Py_ssize_t closure_given)
```

I will need to add assertions for NULL there, although of course easily this could be optimized away by the zig cc.

**Comment by polarathene** [comment]
> Strangely linking didn't work for me any longer, as it also needs a "cc" injection.

This can be necessary if using Zig elsewhere in my experience.

Depends on the environment. Normally `cc` executable is called (_such as by Rust even when `CC` ENV is set it seems_) and that may be installed via `gcc`.
- If pointing that to `zig cc` you would need a wrapper as not all link args are supported, nor are the target values equivalent to GCC/Clang (_Zig is also considering making adjustments here before it's 1.0 release I think_).
- Zig itself will also query `cc`, which if pointing to GCC adds additional search dirs and other paths based on what GCC responds with.

Usually the linker is normally a separate executable that the compiler can call, but Zig has it's own internal linker and it is not flexible to change that (eg via `-fuse-ld=mold`). That's a caveat to keep in mind.

Furthermore, when not using the default `-target native` with Zig, standard system paths will be dropped. You would need to add those back if they're needed (_and compatible with the target you're building for_). You can query those paths like other compilers, `zig cc -target native -print-search-dirs` vs ``zig cc -target x86_64-linux-gnu -print-search-dirs` for example.

---

I still want to share information like above with everything I've written up, but I'm in a rush atm to prepare for an impending power outage event which has taken priority 😅

**Comment by kayhayen** [comment]
In theory, if it works well enough, we could switch from winlibs gcc to download zig on Windows. I am currently making it work on Windows as well, and cleaning up the loose ends to make it work somehow without ccache/clache, which can be disabled. It's a bit annoying since injecting that "cc" is a bit like adding the clcache/ccache binary as a prefix.

**Comment by polarathene** [comment]
> to make it work somehow without ccache

You do not need ccache, I touch on this in my write up but Zig is providing it's own cache that works well.

In a test project Nuitka took the same 6 minutes as with ccache to process modules and generate C source redundantly (re-runs, nothing changed). While without cache it would take approx 30 minutes with building source on a 4 vCPU VPS with 8GB RAM.

---

> It's a bit annoying since injecting that "cc" is a bit like adding the clcache/ccache binary as a prefix.

I know you are trying to be helpful with the Nuitka support for Zig, but at times I wonder if it's making it more problematic than implementing support with typical `CC` ENV that is used with Zig elsewhere via a wrapper script.

It became a slight issue when I needed one specific to Nuitka to workaround issues it presented, while another for compiling with anything else (_python packages via sdists which varied in languages used to compile libs_).

I understand wanting to make it more convenient to use Zig, but I was surprised when Nuitka was detecting Zig via the value of `CC` having `zig` in the path. I have seen some call their wrappers `zcc` and `zxx` for example, so implicit logic like that can be a bit jarring vs just setting `CC` and using `--clang`? (or `--zig`) Less surprises is better 😅

**Comment by kayhayen** [comment]
Yes, I want no extra steps, and calling wrappers is still very expensive, especially on Windows, definitely more expensive than a string operation in Python. 

I was unaware of the default caching behaviour, but that explains how quick it was for me in testing now. We need to force its cache into the Nuitka cache directories though much like it's done with clcache and ccache, and actually I really like to know if the cache is used, so some logging should be added at least as a TODO, and `--disable-ccache` should be made effective, so actual comparisons of compile times can be made.

**Comment by kayhayen** [comment]
Turns out, zig caching cannot be disabled, but be redirected to a folder that you can then delete later on. Now that's strange, as from my experience some kinds of users are really very much into disabling caching for reliability.

**Comment by kayhayen** [comment]
I think there is nothing really that is blocking Nuitka from using zig, so can close this. I am going to track the release of the zig support in #2741 and I am not sure, what to make out of zig bugs we find. Will they accept non-trivial reproducers for zig crashes, like the one I encountered with --debug on Linux?

**Comment by kayhayen** [comment]
So, I guess, we can close this now that zig is kind of baseline. With newer Python, zig will be downloaded on Windows and be the default compiler people use unless they have MSVC installed.
